addon/components/step-manager.js

```diff
@@ -189,25 +232,32 @@ export default Component.extend({
      */
     'transition-to-step'(to, value) {
       const from = get(this, 'transitions.currentStep');
+      const validator = this['will-transition'];
 
-      if (
-        this['will-transition'] &&
-        this['will-transition']({ value, from, to }) === false
-      ) {
+      // Prevent multiple transitions
+      if (get(this, 'loading')) {
         return;
       }
 
-      // Update the `currentStep` if it's mutable
-      if (this.attrs.currentStep && this.attrs.currentStep.update) {
-        this.attrs.currentStep.update(to);
-      }
+      if (validator && typeof validator === 'function') {
+        const result = validator({ value, from, to });
 
-      // Activate the next step
-      get(this, 'transitions').activate(to);
-
-      if (this['did-transition']) {
-        this['did-transition']({ value, from, to });
+        if (result && typeof result.then === 'function') {
```

Comment:
Instead of handling functions that return a Promise and those that do not separately, you could wrap `result` in a promise. If it's already a promise, it'll do nothing, or it'll Promise-ify the result. This leads to less defensive coding, which is generally a good thing.

It would look something like this:

```js
return Promise.resolve(validator({ value, from, to })).then(result => {
  if (result) {
    // Do the transition
  }

  // Otherwise, don't do the transition
});
```

Reply:
Yeah ! I love this piece of code. I will make that change soon after we agree about the Promise usage.

Depending on the outcome, I will either use the code above or this code :
```javascript
var validation = validator({ value, from, to });

if (validation === false) {
  return;
} else {
  Promise.resolve(validation).then(result => {
    // Do the transition
  }, error => {
    // Don't do the transition
  });
}
```